ccl/changefeedccl: TestChangefeedHandlesDrainingNodes failed

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4414439&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4414439&tab=artifacts#/) on master @ [8d0a1773a6b55f06c04ea77e6a39426e32ba9a63](https://github.com/cockroachdb/cockroach/commits/8d0a1773a6b55f06c04ea77e6a39426e32ba9a63):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/a77002d7c9453d7cd2d382f907780e13/logTestChangefeedHandlesDrainingNodes2686424424
    test_log_scope.go:80: use -show-logs to present logs inline
    changefeed_test.go:4427: pq: rangefeeds require the kv.rangefeed.enabled setting. See https://www.cockroachlabs.com/docs/dev/change-data-capture.html#enable-rangefeeds-to-reduce-latency
    panic.go:642: -- test log scope end --
--- FAIL: TestChangefeedHandlesDrainingNodes (7.64s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13280

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4421764&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4421764&tab=artifacts#/) on master @ [1a89121d67fa02d5c674766ea01d3c78455fd28b](https://github.com/cockroachdb/cockroach/commits/1a89121d67fa02d5c674766ea01d3c78455fd28b):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/a77002d7c9453d7cd2d382f907780e13/logTestChangefeedHandlesDrainingNodes3361977065
    test_log_scope.go:80: use -show-logs to present logs inline
    changefeed_test.go:4427: pq: rangefeeds require the kv.rangefeed.enabled setting. See https://www.cockroachlabs.com/docs/dev/change-data-capture.html#enable-rangefeeds-to-reduce-latency
    panic.go:642: -- test log scope end --
--- FAIL: TestChangefeedHandlesDrainingNodes (8.91s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445509&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445509&tab=artifacts#/) on master @ [a7449866f3229872d92e1de973e209838f07de35](https://github.com/cockroachdb/cockroach/commits/a7449866f3229872d92e1de973e209838f07de35):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/a77002d7c9453d7cd2d382f907780e13/logTestChangefeedHandlesDrainingNodes4055445209
    test_log_scope.go:80: use -show-logs to present logs inline
    changefeed_test.go:4489: pq: rangefeeds require the kv.rangefeed.enabled setting. See https://www.cockroachlabs.com/docs/dev/change-data-capture.html#enable-rangefeeds-to-reduce-latency
    panic.go:642: -- test log scope end --
--- FAIL: TestChangefeedHandlesDrainingNodes (12.71s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4488337&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4488337&tab=artifacts#/) on master @ [573b3b05458667308bafeb9db1611832dca3025f](https://github.com/cockroachdb/cockroach/commits/573b3b05458667308bafeb9db1611832dca3025f):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/a77002d7c9453d7cd2d382f907780e13/logTestChangefeedHandlesDrainingNodes157410923
    test_log_scope.go:80: use -show-logs to present logs inline
    changefeed_test.go:4490: pq: rangefeeds require the kv.rangefeed.enabled setting. See https://www.cockroachlabs.com/docs/dev/change-data-capture.html#enable-rangefeeds-to-reduce-latency
    panic.go:642: -- test log scope end --
--- FAIL: TestChangefeedHandlesDrainingNodes (3.57s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -4227,8 +4227,8 @@ func TestChangefeedNodeShutdown(t *testing.T) {
 	defer tc.Stopper().Stop(context.Background())
 
 	db := tc.ServerConn(1)
+	serverutils.SetClusterSetting(t, tc, "changefeed.experimental_poll_interval", time.Millisecond)
 	sqlDB := sqlutils.MakeSQLRunner(db)
-	sqlDB.Exec(t, `SET CLUSTER SETTING changefeed.experimental_poll_interval = '0ns'`)
 	sqlDB.Exec(t, `CREATE DATABASE d`)
 	sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY, b STRING)`)
 	sqlDB.Exec(t, `INSERT INTO foo VALUES (0, 'initial')`)
@@ -4439,19 +4439,22 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 	skip.UnderRace(t, "Takes too long with race enabled")
 
 	shouldDrain := true
-	knobs := base.TestingKnobs{DistSQL: &execinfra.TestingKnobs{
-		DrainFast:  true,
-		Changefeed: &TestingKnobs{},
-		Flowinfra: &flowinfra.TestingKnobs{
-			FlowRegistryDraining: func() bool {
-				if shouldDrain {
-					shouldDrain = false
-					return true
-				}
-				return false
+	knobs := base.TestingKnobs{
+		DistSQL: &execinfra.TestingKnobs{
+			DrainFast:  true,
+			Changefeed: &TestingKnobs{},
+			Flowinfra: &flowinfra.TestingKnobs{
+				FlowRegistryDraining: func() bool {
+					if shouldDrain {
+						shouldDrain = false
+						return true
+					}
+					return false
+				},
 			},
 		},
-	}}
+		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+	}
 
 	sinkDir, cleanupFn := testutils.TempDir(t)
 	defer cleanupFn()
@@ -4466,9 +4469,9 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 
 	db := tc.ServerConn(1)
 	sqlDB := sqlutils.MakeSQLRunner(db)
-	sqlDB.Exec(t, `SET CLUSTER SETTING kv.rangefeed.enabled = true`)
-	sqlDB.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s'`)
-	sqlDB.Exec(t, `SET CLUSTER SETTING changefeed.experimental_poll_interval = '10ms'`)
+	serverutils.SetClusterSetting(t, tc, "kv.rangefeed.enabled", true)
+	serverutils.SetClusterSetting(t, tc, "kv.closed_timestamp.target_duration", time.Second)
+	serverutils.SetClusterSetting(t, tc, "changefeed.experimental_poll_interval", 10*time.Millisecond)
 
 	sqlutils.CreateTable(
 		t, db, "foo",
@@ -4491,9 +4494,9 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 	defer closeFeed(t, feed)
 
 	// At this point, the job created by feed will fail to start running on node 0 due to draining
-	// registry.  However, this job will be retried, and it should succeeded.
+	// registry.  However, this job will be retried, and it should succeed.
 	// Note: This test is a bit unrealistic in that if the registry is draining, that
-	// means that the server is draining (i.e being shut down).  We don't do a full shutdown
+	// means that the server is draining (i.e. being shut down).  We don't do a full shutdown
 	// here, but we are simulating a restart by failing to start a flow the first time around.
 	assertPayloads(t, feed, []string{
 		`foo: [1]->{"after": {"k": 1, "v": 1}}`,

--- a/pkg/testutils/serverutils/BUILD.bazel
+++ b/pkg/testutils/serverutils/BUILD.bazel
@@ -4,6 +4,7 @@ go_library(
     name = "serverutils",
     srcs = [
         "test_cluster_shim.go",
+        "test_cluster_utils.go",
         "test_server_shim.go",
         "test_tenant_shim.go",
     ],
@@ -20,6 +21,7 @@ go_library(
         "//pkg/server/status",
         "//pkg/settings/cluster",
         "//pkg/storage",
+        "//pkg/testutils",
         "//pkg/testutils/sqlutils",
         "//pkg/util/hlc",
         "//pkg/util/httputil",

--- a/pkg/testutils/serverutils/test_cluster_utils.go
+++ b/pkg/testutils/serverutils/test_cluster_utils.go
@@ -1,0 +1,50 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package serverutils
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+)
+
+// SetClusterSetting executes set cluster settings statement, and then ensures that
+// all nodes in the test cluster see that setting update.
+func SetClusterSetting(t testutils.TB, c TestClusterInterface, name string, value interface{}) {
+	t.Helper()
+	strVal := func() string {
+		switch v := value.(type) {
+		case string:
+			return v
+		case int, int32, int64:
+			return fmt.Sprintf("%d", v)
+		case bool:
+			return strconv.FormatBool(v)
+		case float32, float64:
+			return fmt.Sprintf("%f", v)
+		case fmt.Stringer:
+			return v.String()
+		default:
+			return fmt.Sprintf("%v", value)
+		}
+	}()
+	query := fmt.Sprintf("SET CLUSTER SETTING %s='%s'", name, strVal)
+	// Set cluster setting statement ensures the setting is propagated to the local registry.
+	// So, just execute the query against each node in the cluster.
+	for i := 0; i < c.NumServers(); i++ {
+		_, err := c.ServerConn(i).ExecContext(context.Background(), query)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}

